### DISCUSSION TITLE
WebUpdate Show Firmware Version

### DISCUSSION THREAD

**Discussion by lucho512** [Hieromon/AutoConnect#343] [Ideas]
As I can add a label in the webupdate page that indicates the firmware version, this variable would be defined in the main.

Thanks!!!

**Comment by Hieromon** [comment]
@lucho512 
> As I can add a label in the webupdate page that indicates the firmware version,

How?, Which the webupdate page?, Can you offer a concrete way of doing that?

**Comment by lucho512** [comment]
Hi Hieromon,

How can I show the current firmware version on the update page? the idea would be to add an actext in the update page to send the value that contains the firmware version

I hope I have explained myself, my English is poor

Greetings

**Comment by Hieromon** [comment]
Your intention is to be feedback the variable value indicating the version number defined in the sketch to the update page when you select the firmware that can be updated by [AutoConnectOTA](https://hieromon.github.io/AutoConnect/otabrowser.html#updates-with-the-web-browserupdated-wv115).

However, the behavior is inconsistent. By the time AutoConnect displays the update page, AutoConnect itself must have the binary code for the firmware of your choice. This means that you have uploaded the firmware from your browser (which is the client from AutoConnect's perspective). You need to understand how HTTP requests-responses work in a client/server model.

In addition, the version number you embed in your sketch is usually not identifiable within the compiled binary. Strictly identifiable, even the ESP32 lacks resources to incorporate such reverse engineering processes.

A practical technique you can choose is to include an ID to identify the version as part of the compiled BIN file name. 

**NOTE**
This topic is not an issue with the AutoConnect library, so I close it and convert it to Discussions.

**Comment by dncex** [comment]
I don't know if I have correclty understood, but if you just want to quickly show the current firmware running in your board you can simple add this before starting the portal:
config.title = "Autoconnect - V3.33";
This will add the version string in the menu title of the captive portal (replacing V3.33 with a proper constant previosly defined)

**Reply by Hieromon** [reply]
Is your request to show the currently running version? Or do you want to see the target firmware version when you select a file using the *AutoConnectOTA* page (it's not yet started uploading)?

**Reply by dncex** [reply]
no... it was a suggestion for the @lucho512; if I correctly understood the question, he was looking for a way to show in the captive portal the current firmware. With my suggestion it's showing the version in every page.

**Reply by DanMan32** [reply]
I have the same question.   I'd like to put the current version string defined at the beginning of my sketch as a constant and put it somewhere in the Update page by overriding one of the existing text elements in the update page, such as suffixing the caption "Updating Firmware" with " - v"+version.
You'd think a constant would be a problem since the constant is fixed in the binary, but it doesn't change until the firmware is updated and the ESP32 rebooted.

This would be a good example of how to override elements in any of the included Autoconnect menu pages.

**Reply by DanMan32** [reply]
PS. I suffixed the version to menu title for my custom page.   Ugggly!
Menu item 'Solar Pool" became "Solar Pool - v071321-1dev".   That would be fine somewhere in the body of a page, but not that long string in the menu/title bar.
I suppose it would work if I could change the font size mid-string to maybe 1/4 its original size.
I've got a LOT to learn about HTML code in general, let alone injecting it with the Autoconnect element coding.

**Reply by Hieromon** [reply]
@DanMan32 
>  I'd like to put the current version string defined at the beginning of my sketch as a constant and put it somewhere in the Update page by overriding one of the existing text elements in the update page, such as suffixing the caption "Updating Firmware"

Ok, I consider improvements to achieve this.

**Comment by Hieromon** [comment]
@DanMan32
Before I enhance AutoConnect, I'll show you how it can be implemented in the current release. This method applies the `mylabels.h`, which is described in the [documentation](https://hieromon.github.io/AutoConnect/changelabel.html#configuration-of-platformioini). However, it is needed PlatformIO for your build environment.

1. Prepare overriding context

    The "Updating firmware" string can be overwritten with an alternative string when building the sketch. Type the following code in your text editor and save it as `mylabels.h` in a folder at the same level as your sketch.

    ```cpp
    #define MY_VERSION  "v071321-1dev"
    #undef AUTOCONNECT_TEXT_UPDATINGFIRMWARE
    #define AUTOCONNECT_TEXT_UPDATINGFIRMWARE "Updating firmware<span style='display:block;text-align:right;font-size:small'>" MY_VERSION "</span>"
    ```

2. Configuration the `platformio.ini`

    Configure `platormio.ini` file as described [**here**](https://hieromon.github.io/AutoConnect/changelabel.html#configuration-of-platformioini) in the documentation.

3. Build with PlatformIO and replace `AUTOCONNECT_TEXT_UPDATING_FIRMWARE`

    You build with PlatformIO using the above configuration, the string defined by `MY_VERSION` will appear at the right side of the **`Updating firmware`** on the OTA screen.

Note that you need to delete the compiled object each time you update `mylabels.h`.
